### PR TITLE
Continuous Correctness Testing

### PR DESCRIPTION
**Pull request by kennyzzhang**
There's still an issue with the supernode copying. From reading the current code, I was under the impression that I could memcpy the sketch_buffer to another place, then after running connected components, free the used supernodes and point the graph to the copy. But evidently somewhere somethings goes wrong, and it segfaults after the first correctness check.

### CODE REVIEW

**Comment by etwest** [comment]
This is an issue we’ve dealt with before. Part of the supernode/sketch structure is a pointer to where the associated data lives. Your copy constructor will need to be a little more intelligent that a direct memcpy. I can write some code for you if you want.

**Comment by kennyzzhang** [comment]
> 
> 
> This is an issue we’ve dealt with before. Part of the supernode/sketch structure is a pointer to where the associated data lives. Your copy constructor will need to be a little more intelligent that a direct memcpy. I can write some code for you if you want.

Ah okay, I see the pointers, I'll fix that.

File: CMakeLists.txt [hunk 1]
```diff
@@ -354,10 +358,43 @@ else ()
   endif()
 endif ()
 
+add_executable(cont_test
```
Comment (etwest) [inline CMakeLists.txt:356]: So as mentioned earlier today we may want to separate these changes out into just the new verifier stuff and then a separate repo which is for the experiment code

File: test/cont_test.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#include <gtest/gtest.h>
```
Comment (etwest) [inline test/cont_test.cpp:1]: This would be the code that would go into the separate repo along with the cmake stuff.

File: test/cont_test.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  unsigned long num_failure = 0;
+  for (unsigned long i = 0; i < samples; i++) {
+    std::cout << "Starting updates" << std::endl;
+    for (unsigned long j = 0; j < updates_per_sample; j++) {
+      unsigned edgei = rand() % nodes;
+      unsigned edgej = rand() % (nodes - 1);
+      if (edgei > edgej) {
+        std::swap(edgei, edgej);
+      } else {
+        edgej++;
```
Comment (etwest) [inline test/cont_test.cpp:22]: Could there be some weird edge case here where we do an integer overflow and end up with a self loop?
I think so. It's amazingly unlikely but you could just add an else if just in case.

File: test/cont_test.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  }
+  std::clog << n << ',' << num_failure << std::endl;
+}
+
+//TEST(TestContinuous, TestRandom) {
+//  test_continuous(1e5, 1e7, 10);
+//}
+
+TEST(TestContinuous, DISABLED_StandardKron17) {
```
Comment (etwest) [inline test/cont_test.cpp:86]: DISABLED_StandardKron17 is not very specific. Maybe NoCont_Kron17 or something?
Also, why is this here?

File: test/graph_test.cpp [hunk 5]
```diff
@@ -38,15 +39,26 @@ TEST(GraphTestSuite, IFconnectedComponentsAlgRunTHENupdateLocked) {
...
+  g.set_verifier(std::make_unique<FileGraphVerifier>(curr_dir + "/res/multiples_graph_1024.txt"));
   g.connected_components();
   ASSERT_THROW(g.update({{1,2}, INSERT}), UpdateLockedException);
   ASSERT_THROW(g.update({{1,2}, DELETE}), UpdateLockedException);
 }
 
 TEST(GraphTestSuite, TestRandomGraphGeneration) {
   generate_stream();
-  GraphVerifier graphVerifier {};
+  ifstream in { "./cumul_sample.txt" };
```
Comment (etwest) [inline test/graph_test.cpp]: What exactly is this test accomplishing? Just asserting that creating a graph verifier doesn't crash?

File: test/cont_test.cpp (hunk 3 above)
Comment (etwest) [reply test/cont_test.cpp:22]: I guess this isn't possible assuming nodes < unsigned max so never mind

File: CMakeLists.txt (hunk 1 above)
Comment (Victor-C-Zhang) [reply CMakeLists.txt:356]: we can do that once the changes are into main

File: test/cont_test.cpp (hunk 4 above)
Comment (Victor-C-Zhang) [reply test/cont_test.cpp:86]: disabled means the test is skipped. why is it skipped?

File: CMakeLists.txt [hunk 6]
```diff
@@ -4,7 +4,7 @@
 cmake_minimum_required(VERSION 3.15)
 include (ExternalProject)
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 14)
```
Comment (Victor-C-Zhang) [inline CMakeLists.txt:7]: what functionality do we need from C++14?

File: CMakeLists.txt [hunk 7]
```diff
@@ -179,14 +179,14 @@ link_directories(${CMAKE_CURRENT_BINARY_DIR}/BufferTree/prefix/lib)
...
 
-
 add_executable(tests
   test/test_runner.cpp
   test/graph_test.cpp
   test/sketch_test.cpp
   test/supernode_test.cpp
   test/util_test.cpp
-  test/util/graph_verifier.cpp
+  test/util/file_graph_verifier.cpp
```
Comment (Victor-C-Zhang) [inline CMakeLists.txt:188]: why are all these executables using both file and matrix graph verifiers? shouldn't we just pick one?

File: src/graph.cpp [hunk 8]
```diff
@@ -269,11 +295,6 @@ vector<set<node_t>> Graph::parallel_connected_components() {
   return retval;
 }
 
-void Graph::post_cc_resume() {
-  GraphWorker::unpause_workers();
-  update_locked = false;
-}
-
```
Comment (Victor-C-Zhang) [inline src/graph.cpp:289]: @etwest what's the purpose of this function? I don't see it being used in the active codebase, which is probably why Kenny's getting rid of it. At the same time, this seems like something that may have been important.

File: test/cont_test.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  int type, a, b;
+  while (m--) {
+    in >> type >> a >> b;
+    g.update({{a, b}, INSERT});
+  }
+  g.connected_components();
+}
+
+TEST(TestContinuous, TestKron17) {
+  std::ifstream input { "./kron17" };
```
Comment (Victor-C-Zhang) [inline test/cont_test.cpp:101]: maybe add a note reminding that we're assuming we have the kron graph in the runtime directory

File: test/graph_test.cpp (hunk 5 above)
Comment (Victor-C-Zhang) [reply test/graph_test.cpp]: A similar comment: this should probably go into a `*_graph_verifier_test` file.

File: test/util/mat_graph_verifier.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  if (parent[i] == i) return i;
+  return parent[i] = dsu_find(parent[i], parent);
+}
+
+static void dsu_union(node_t i, node_t j, node_t* parent, node_t* size) {
+  i = dsu_find(i, parent); j = dsu_find(j, parent);
+  if (size[i] < size[j]) std::swap(i,j);
+  parent[j] = i;
+  size[i] += size[j];
+}
```
Comment (Victor-C-Zhang) [inline test/util/mat_graph_verifier.cpp]: nit: if this is identical to the `file_graph_verifier` functions, you could probably refactor both of them into the superclass instead.

File: test/util/file_graph_verifier.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+#include <set>
+#include "../../include/supernode.h"
+#include "graph_verifier.h"
+
+/**
+ * A plugin for the Graph class that runs Boruvka alongside the graph algorithm
+ * and verifies the edges and connected components that the graph algorithm
+ * generates.
+ */
```
Comment (Victor-C-Zhang) [inline test/util/file_graph_verifier.h]: Describe what's different between this and matrix verifier

File: test/util/mat_graph_verifier.h [hunk 12]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+#include <set>
+#include "../../include/supernode.h"
+#include "graph_verifier.h"
+
+/**
+ * A plugin for the Graph class that runs Boruvka alongside the graph algorithm
+ * and verifies the edges and connected components that the graph algorithm
+ * generates.
+ */
```
Comment (Victor-C-Zhang) [inline test/util/mat_graph_verifier.h]: Describe what's different between this and file verifier

File: src/graph.cpp (hunk 8 above)
Comment (etwest) [reply src/graph.cpp:289]: I think it’s important to keep. Kenny is getting rid of it because his code is implementing its functionality itself. However if we write other dynamic querying code then we can use it.

maybe we can use it in Kenny’s functions so it shows up somewhere at least.

File: CMakeLists.txt (hunk 6 above)
Comment (kennyzzhang) [reply CMakeLists.txt:7]: std::make_unique, if we want to use smart pointers it's nice to have. If c++14 is too new a standard, I can rewrite it in c++11.

File: src/graph.cpp (hunk 8 above)
Comment (kennyzzhang) [reply src/graph.cpp:289]: It feels like the only reason we'd need to do the `update_locked = false` operation is when we're copying the supernodes to allow requerying, which is technically not allowed by theory. The only reason we implemented this function was to do the continuous testing anyways.

File: test/cont_test.cpp (hunk 4 above)
Comment (kennyzzhang) [reply test/cont_test.cpp:86]: I didn't write this test, and I have no idea why it's here. It looks like it's just doing a normal graph verification on kron17. Shall we just remove it?

>```
> 1ce95873 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:37:57 -0700  86) TEST(TestContinuous, DISABLED_StandardKron17) {
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  87)   std::ifstream in{ "./kron17" };
> 36fa97dc test/cont_test.cpp                   (kennyzzhang       2021-10-17 16:28:20 -0400  88)   node_t n, m;
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  89)   in >> n >> m;
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  90)   Graph g{n};
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  91)   int type, a, b;
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  92)   while (m--) {
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  93)     in >> type >> a >> b;
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  94)     g.update({{a, b}, INSERT});
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  95)   }
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  96)   g.connected_components();
> cc04cd18 test/cont_test.cpp                   (Victor Zhang      2021-08-15 23:08:44 -0700  97) }
>```

File: test/graph_test.cpp (hunk 5 above)
Comment (kennyzzhang) [reply test/graph_test.cpp]: I have no idea what this test is accomplishing, so I'm getting rid of it.

File: src/graph.cpp (hunk 8 above)
Comment (etwest) [reply src/graph.cpp:289]: Future work will be in dynamic querying. For which having the ability to lock/unlock updates and pause/resume graph_workers will be essential

Comment (kennyzzhang) [reply src/graph.cpp:289]: For dynamic querying, pass `true` to connected_components. The code has only been moved to a place where a user can't screw up resuming connected components without creating a backup copy of the supernodes.

> ```
> vector<set<node_t>> Graph::connected_components(bool cont) {
>   if (!cont)
>     return connected_components();
> 
> #ifdef USE_FBT_F
>   bf->force_flush(); // flush everything in buffertree to make final updates
> #else
>   wq->force_flush();
> #endif
> 
>   // Copy supernodes
>   Supernode** supernodes = new Supernode*[num_nodes];
>   for (node_t i=0;i<num_nodes;++i) {
>     supernodes[i] = Supernode::makeSupernode(*this->supernodes[i]);
>   }
> 
>   vector<set<node_t>> ret;
>   ret = connected_components();
>   // Restore supernodes
>   for (node_t i=0;i<num_nodes;++i) {
>     free(this->supernodes[i]);
>     this->supernodes[i] = supernodes[i];
>     representatives->insert(i);
>     parent[i] = i;
>   }
>   delete[] supernodes;
> 
>   GraphWorker::unpause_workers();
>   update_locked = false;
>   return ret;
> }
> ```

File: src/graph.cpp [hunk 13]
```diff
@@ -180,12 +177,44 @@ vector<set<node_t>> Graph::connected_components() {
...
+  if (!cont)
+    return connected_components();
+
+#ifdef USE_FBT_F
+  bf->force_flush(); // flush everything in buffertree to make final updates
+#else
+  wq->force_flush();
+#endif
+
+  // Copy supernodes
```
Comment (etwest) [inline src/graph.cpp:194]: So if you don't pause the graph workers before copying the supernodes there's no guarantee that the super nodes have the right state. They could have too few updates applied to them. This is a rare edge case but it does exist here.

File: test/cont_test.cpp (hunk 4 above)
Comment (etwest) [reply test/cont_test.cpp:86]: Sure thing

File: src/graph.cpp (hunk 8 above)
Comment (etwest) [reply src/graph.cpp:289]: Well this is probably not the algorithm we'll use so that complicates things a little bit. Personally, I think its cleaner to have but its also not a big deal not to have it. It's only two lines of code and can be reintroduced as necessary.

File: CMakeLists.txt (hunk 6 above)
Comment (Victor-C-Zhang) [reply CMakeLists.txt:7]: I recall in a design discussion we had early on that we want to stick with the 11 standard

Comment (etwest) [reply CMakeLists.txt:7]: Conclusion: We decided to go with c++14

**Review by etwest — APPROVED** [review]
LGTM
